File: Game Autosaver/GameList.vb
Imports Newtonsoft.Json

Public Class GameList

    Private Sub GameList_Load() Handles MyBase.Load
        Icon = My.Resources.Icon_256

        MainForm.Games.SaveCurrentSettings(MainForm.ThisForm)

        If MainForm.Games.GameList.Count > 0 Then
            For Each e As GameSettings In MainForm.Games.GameList.Values
                Dim NewRowIndex As Integer = DataGridView1.Rows.Add(e.Name, "Load", "Remove")

                If e.Name = MainForm.Label6.Text Then ' currently open save
                    Dim CellBtn As New DataGridViewButtonCell
                    CellBtn.Value = "Current"
                    DataGridView1.Item(1, NewRowIndex) = CellBtn
                End If
                DataGridView1.Rows(NewRowIndex).Selected = False
            Next
            DataGridView1.Sort(DataGridView1.Columns.Item(0), System.ComponentModel.ListSortDirection.Ascending)
        End If

    End Sub

    Private Sub GameList_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing

    End Sub

    Private Sub DataGridView1_CellClick(sender As Object, e As DataGridViewCellEventArgs) Handles DataGridView1.CellClick
        If e.RowIndex > -1 Then
            Select Case e.ColumnIndex
                Case 1 ' Load Button
                    LoadGame(e.RowIndex)
                Case 2 ' Remove Button
                    RemoveGame(e.RowIndex)
            End Select
        End If
    End Sub

    ''' <summary>
    ''' Load a corresponding row in the DataGridView.
    ''' </summary>
    ''' <param name="row">The corresponding row.</param>
    Private Sub LoadGame(row As Integer)
        If row >= 0 Then
            If CStr(DataGridView1.Item(1, row).Value) = "Current" Then
                MsgBox("This is already active.", MsgBoxStyle.Information, "Error")
            Else
                If DataGridView1.SelectedCells.Count > 0 Then

                    Dim idx As Integer = DataGridView1.SelectedCells.Item(0).RowIndex
                    Dim Name As String = CStr(DataGridView1.Item(0, idx).Value)
                    If MainForm.Games.GameList.ContainsKey(Name) Then
                        StopAnyMainFormWork() ' stop any active elements
                        MainForm.MainForm_FormClosing() ' update settings

                        If MainForm.Games.LoadSave(MainForm.ThisForm, Name) Then
                            My.Settings.Saves = JsonConvert.SerializeObject(MainForm.Games)
                            My.Settings.Save()
                            MainForm.TopMost = True
                            MainForm.TopMost = False
                            Me.Close()
                        Else
                            MsgBox("Could not find game data.", MsgBoxStyle.Exclamation, "Error")
                        End If
                    Else
                        MsgBox("Could not find saved game with that name.", MsgBoxStyle.Exclamation, "Error")
                    End If

                End If
            End If
        End If
    End Sub

    Private Sub ResortDataGridView()
        If DataGridView1.SortOrder <> SortOrder.None AndAlso DataGridView1.SortedColumn IsNot Nothing Then
            Dim Direction As System.ComponentModel.ListSortDirection = System.ComponentModel.ListSortDirection.Ascending
            If DataGridView1.SortOrder = SortOrder.Descending Then
                Direction = System.ComponentModel.ListSortDirection.Descending
            End If
            DataGridView1.Sort(DataGridView1.SortedColumn, Direction)
        End If
    End Sub

    ''' <summary>
    ''' Stop any active elements in the MainForm.
    ''' </summary>
    Private Shared Sub StopAnyMainFormWork()
        SyncLock MainForm.accessLock
            If MainForm.StartButton.Text = "Stop Autosaving" Then
                MainForm.StartButton.PerformClick() ' stop autosaving
            End If
        End SyncLock
        SettingsForm.Close()
        BackgroundImageDialog.Close()
        SavesViewer.Close()
    End Sub

    ''' <summary>
    ''' Remove a specified row and all it's data from settings.
    ''' </summary>
    ''' <param name="row">Corresponding row. (Name and settings data row)</param>
    Private Sub RemoveGame(row As Integer)
        If row >= 0 Then
            If DataGridView1.Rows.Count > 0 Then
                If CStr(DataGridView1.Item(1, row).Value) = "Current" Then
                    MsgBox("This is currently active so it can't be deleted.", MsgBoxStyle.Information, "Error")
                Else
                    My.Computer.Audio.PlaySystemSound(System.Media.SystemSounds.Exclamation)
                    Dim result As MsgBoxResult = MsgBox("Are you sure you want to delete?", MsgBoxStyle.YesNo, "Warning")
                    If result = MsgBoxResult.Yes Then
                        If MainForm.Games.GameList.Remove(CStr(DataGridView1.Item(0, row).Value)) Then
                            DataGridView1.Rows.RemoveAt(row)
                        Else
                            MsgBox("Could not find data to delete!", MsgBoxStyle.Critical, "Error")
                        End If
                    End If
                End If
            End If
        End If
    End Sub

    Public Const GameNameCharLimit As Integer = 300
    ''' <summary>
    ''' Create and add a new row to the DataGridView
    ''' </summary>
    Private Sub NewButton_Click(sender As Object, e As EventArgs) Handles NewButton.Click
        Dim NewGame As String = InputBox("Enter a name", "New", "")

        If NewGame = "" Then
            'My.Computer.Audio.PlaySystemSound(System.Media.SystemSounds.Exclamation)
        ElseIf NewGame.Length >= GameNameCharLimit Then
            MsgBox("Name cannot be longer than " & GameNameCharLimit & ".", MsgBoxStyle.Exclamation, "Error")
        Else
            For i As Integer = 0 To DataGridView1.Rows.Count - 1
                If CStr(DataGridView1.Item(0, i).Value) = NewGame Then
                    MsgBox("This name already exists!", MsgBoxStyle.Exclamation, "Error")
                    Exit Sub
                End If
            Next

            MainForm.Games.GameList.Add(NewGame, New GameSettings)
            MainForm.Games.GameList(NewGame).Name = NewGame
            DataGridView1.Rows.Add(NewGame, "Load", "Remove")
            ResortDataGridView()
        End If
    End Sub

    Private Sub DataGridView1_CellValueChanged(sender As Object, e As DataGridViewCellEventArgs) Handles DataGridView1.CellValueChanged

        ' renaming row
        If e.ColumnIndex = 0 AndAlso DataGridView1.SelectedCells.Count > 0 Then
            Dim NewName As String = CStr(DataGridView1.Item(e.ColumnIndex, EdittedRow).Value)
            If NewName <> "" Then
                If MainForm.Games.RenameGameName(MainForm.ThisForm, OldName, NewName) Then
                    MainForm.SaveMySettings()
                Else
                    MsgBox("Name already exists.", MsgBoxStyle.Exclamation)
                End If
            End If
        End If
    End Sub

    Private EdittedRow As Integer = -1
    Private OldName As String = ""
    Private Sub DataGridView1_CellBeginEdit(sender As Object, e As DataGridViewCellCancelEventArgs) Handles DataGridView1.CellBeginEdit
        EdittedRow = e.RowIndex
        OldName = CStr(DataGridView1.Item(0, e.RowIndex).Value)
    End Sub
End Class

File: Game Autosaver/GameList.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Newtonsoft.Json;
using Microsoft.VisualBasic;
using System.Media;

namespace GameAutosaver
{
    public partial class GameList : Form
    {
        public GameList(MainForm form)
        {
            InitializeComponent();
            mainForm = form;
        }

        MainForm mainForm;

        private void GameList_Load(object sender, EventArgs e1)
        {
            Icon = Properties.Resources.Icon_256;

            MainForm.Games.SaveCurrentSettings(MainForm.thisForm);

            if (MainForm.Games.GameList.Count > 0) {
                foreach (GameSettings e in MainForm.Games.GameList.Values) {
                    int NewRowIndex = DataGridView1.Rows.Add(e.Name, "Load", "Remove");

                    if (e.Name == mainForm.Label6.Text) { // currently open save
                        DataGridViewButtonCell CellBtn = new DataGridViewButtonCell();
                        CellBtn.Value = "Current";
                        DataGridView1[1, NewRowIndex] = CellBtn;
                    }
                    DataGridView1.Rows[NewRowIndex].Selected = false;
                }
                DataGridView1.Sort(DataGridView1.Columns[0], System.ComponentModel.ListSortDirection.Ascending);
            }
        }

        private void GameList_FormClosing(object sender, FormClosingEventArgs e)
        {

        }

        private void DataGridView1_CellClick(object sender, DataGridViewCellEventArgs e)
        {
            if (e.RowIndex > -1) {
                switch (e.ColumnIndex) {
                    case 1: // Load Button
                        LoadGame(e.RowIndex);
                        break;
                    case 2: // Remove Button
                        RemoveGame(e.RowIndex);
                        break;
                }
            }
        }

        /// <summary>
        /// Load a corresponding row in the DataGridView.
        /// </summary>
        /// <param name="row">The corresponding row.</param>
        private void LoadGame(int row, bool dontSaveGameBeforeLoad = false)
        {
            if (row >= 0) {
                if (Convert.ToString(DataGridView1[1, row].Value) == "Current") {
                    Interaction.MsgBox("This is already active.", MsgBoxStyle.Information, "Error");
                } else {

                    string Name = Convert.ToString(DataGridView1[0, row].Value);
                    if (MainForm.Games.GameList.ContainsKey(Name)) {
                        StopAnyMainFormWork(); // stop any active elements

                        if (MainForm.Games.LoadSave(MainForm.thisForm, Name, dontSaveGameBeforeLoad)) {
                            mainForm.MainForm_FormClosing(); // update settings
                            mainForm.Activate();
                            this.Close();
                        } else {
                            Interaction.MsgBox("Could not find game data.", MsgBoxStyle.Exclamation, "Error");
                        }
                    } else {
                        Interaction.MsgBox("Could not find saved game with that name.", MsgBoxStyle.Exclamation, "Error");
                    }

                }
            }
        }


        private void ResortDataGridView()
        {
            if (DataGridView1.SortOrder != SortOrder.None && DataGridView1.SortedColumn != null) {
                System.ComponentModel.ListSortDirection Direction = System.ComponentModel.ListSortDirection.Ascending;
                if (DataGridView1.SortOrder == SortOrder.Descending) {
                    Direction = System.ComponentModel.ListSortDirection.Descending;
                }
                DataGridView1.Sort(DataGridView1.SortedColumn, Direction);
            }
        }

        /// <summary>
        /// Stop any active elements in the MainForm.
        /// </summary>
        private static void StopAnyMainFormWork()
        {
            lock (MainForm.thisForm.accessLock) {
                if (MainForm.thisForm.StartButton.Text == "Stop Autosaving") {
                    MainForm.thisForm.StartButton.PerformClick(); // stop autosaving
                }
            }
            if (MainForm.settingsForm != null && MainForm.settingsForm.Visible)
                MainForm.settingsForm.Close();
            if (SettingsForm.backgroundImageDialog != null && SettingsForm.backgroundImageDialog.Visible)
                SettingsForm.backgroundImageDialog.Close();
            if (MainForm.savesViewer != null && MainForm.savesViewer.Visible)
                MainForm.savesViewer.Close();
        }

        /// <summary>
        /// Remove a specified row and all it's data from settings.
        /// </summary>
        /// <param name="row">Corresponding row. (Name and settings data row)</param>
        private void RemoveGame(int row)
        {
            if (row >= 0) {
                if (DataGridView1.Rows.Count > 0) {
                    SystemSounds.Exclamation.Play();
                    MsgBoxResult result = Interaction.MsgBox("Are you sure you want to delete?", MsgBoxStyle.YesNo, "Warning");
                    if (result == MsgBoxResult.Yes) {
                        StopAnyMainFormWork(); // stop any active elements
                        if (MainForm.Games.GameList.Remove(Convert.ToString(DataGridView1[0, row].Value))) {
                            string txt = Convert.ToString(DataGridView1[1, row].Value);
                            DataGridView1.Rows.RemoveAt(row);
                            if (txt == "Current") {
                                const string name = "GAME AUTOSAVER";
                                bool found = false;
                                for (int i = 0; i <= DataGridView1.Rows.Count - 1; i++) {
                                    if (Convert.ToString(DataGridView1[0, i].Value) == name) {
                                        LoadGame(i, true);
                                        found = true;
                                    }
                                }
                                if (!found) {
                                    MainForm.Games.GameList.Add(name, new GameSettings());
                                    MainForm.Games.GameList[name].Name = name;
                                    int idx = DataGridView1.Rows.Add(name, "Load", "Remove");
                                    //ResortDataGridView();
                                    LoadGame(idx, true);
                                }
                            }
                        } else {
                            Interaction.MsgBox("Could not find data to delete!", MsgBoxStyle.Critical, "Error");
                        }
                    }
                }
            }
        }

        public const int GameNameCharLimit = 300;
        /// <summary>
        /// Create and add a new row to the DataGridView
        /// </summary>
        private void NewButton_Click(object sender, EventArgs e)
        {
            string NewGame = Interaction.InputBox("Enter a name", "New", "");

            if (string.IsNullOrEmpty(NewGame)) {
                //SystemSounds.Exclamation.Play();
            } else if (NewGame.Length >= GameNameCharLimit) {
                Interaction.MsgBox("Name cannot be longer than " + GameNameCharLimit + ".", MsgBoxStyle.Exclamation, "Error");
            } else {
                for (int i = 0; i <= DataGridView1.Rows.Count - 1; i++) {
                    if (Convert.ToString(DataGridView1[0, i].Value) == NewGame) {
                        Interaction.MsgBox("This name already exists!", MsgBoxStyle.Exclamation, "Error");
                        return;
                    }
                }

                MainForm.Games.GameList.Add(NewGame, new GameSettings());
                MainForm.Games.GameList[NewGame].Name = NewGame;
                DataGridView1.Rows.Add(NewGame, "Load", "Remove");
                ResortDataGridView();
            }
        }


        private void DataGridView1_CellValueChanged(object sender, DataGridViewCellEventArgs e)
        {
            // renaming row
            if (e.ColumnIndex == 0 && DataGridView1.SelectedCells.Count > 0) {
                string NewName = Convert.ToString(DataGridView1[e.ColumnIndex, EdittedRow].Value);
                if (!string.IsNullOrEmpty(NewName)) {
                    if (MainForm.Games.RenameGameName(MainForm.thisForm, OldName, NewName)) {
                        mainForm.SaveMySettings();
                    } else {
                        Interaction.MsgBox("Name already exists.", MsgBoxStyle.Exclamation);
                    }
                }
            }
        }

        private int EdittedRow = -1;
        private string OldName = "";
        private void DataGridView1_CellBeginEdit(object sender, DataGridViewCellCancelEventArgs e)
        {
            EdittedRow = e.RowIndex;
            OldName = Convert.ToString(DataGridView1[0, e.RowIndex].Value);
        }
    }
}
